fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2AnalyticRule.java

```diff
@@ -0,0 +1,869 @@
+                    .filter(LogicalJoinOperator.class::isInstance)
+                    .map(Operator::<LogicalJoinOperator>cast)
+                    .allMatch(joinOp -> joinOp.getJoinType().isCrossJoin());
+            boolean subqueryAllCrossJoin = rightOps.stream()
+                    .filter(LogicalJoinOperator.class::isInstance)
+                    .map(Operator::<LogicalJoinOperator>cast)
+                    .allMatch(joinOp -> joinOp.getJoinType().isCrossJoin());
+
+            return outerAllCrossJoin && subqueryAllCrossJoin;
+        }
+
+        /**
+         * The following requirements should have been met
+         * 1. No other aggregate node allowed in the subquery
+         * 2. All the aggregate functions must have corresponding analytic version,
+         * which is defined in TRANSFORMABLE_AGGREGATE_FUNCTIONS
+         * 3. All the aggregate shouldn't be distinct
+         */
+        private boolean checkAggregate() {
+            if (rightOps.stream()
+                    .filter(LogicalAggregationOperator.class::isInstance)
+                    .count() > 1) {
+                return false;
+            }
+
+            List<CallOperator> aggregations = Lists.newArrayList(subAggOp.getAggregations().values());
+            if (aggregations.stream()
+                    .anyMatch(callOp -> !TRANSFORMABLE_AGGREGATE_FUNCTIONS.contains(callOp.getFnName()))) {
+                return false;
+            }
+
+            return aggregations.stream().noneMatch(CallOperator::isDistinct);
+        }
+
+        /**
+         * Check whether `outerTables = subqueryTables + correlatedOuterTable` satisfied
+         * The above requirement must be required, otherwise subquery cannot be eliminated through window function
+         */
+        private boolean checkReferencedTables() {
+            List<Table> outerTables = getAllTables(leftOps);
+            List<Table> subqueryTables = getAllTables(rightOps);
+            outerColumnRefOperators = getAllColumnRefOperators(leftOps);
+            List<Column> correlatedOuterColumns = applyOp.getCorrelationColumnRefs().stream()
+                    .map(context.getColumnRefFactory()::getColumn)
+                    .collect(Collectors.toList());
+
+            Map<Long, Table> outerTableMap = outerTables.stream()
+                    .collect(Collectors.toMap(Table::getId, Function.identity()));
+            if (!subqueryTables.stream()
+                    .map(Table::getId)
+                    .allMatch(outerTableMap::containsKey)) {
+                return false;
+            }
+            subqueryTables.forEach(table -> outerTableMap.remove(table.getId()));
+            if (outerTableMap.size() != 1) {
+                return false;
+            }
+            correlatedOuterTable = outerTableMap.values()
+                    .stream()
+                    .findFirst()
+                    .orElseThrow(NoSuchElementException::new);
```

Comment:
could we throw exception here?

Reply:
Exception will never be throwed here, I just don't want to check `Optional.isPresent`